File: C#Solutions/problem23.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;

namespace AllProblems
{
    public static class Problem23
    {
        public class ListNode : IComparable<ListNode>
        {
            public int val;
            public ListNode next;

            public ListNode(int val = 0, ListNode next = null)
            {
                this.val = val;
                this.next = next;
            }

            public int CompareTo(ListNode other)
            {
                return this.val.CompareTo(other.val);
            }

            public override string ToString()
            {
                return val + (next != null ? " -> " + next.ToString() : " -> null");
            }
        }

        // Method to merge k sorted linked lists
        public static ListNode MergeKSortedLists(List<ListNode> lists)
        {
            if (lists == null || lists.Count == 0)
                return null;

            // Use a priority queue to handle duplicates
            var priorityQueue = new PriorityQueue<ListNode, int>();

            // Add the head of each list to the priority queue
            foreach (var list in lists)
            {
                if (list != null)
                {
                    priorityQueue.Enqueue(list, list.val);
                }
            }

            ListNode dummy = new ListNode();
            ListNode current = dummy;

            // Extract nodes from the priority queue and build the merged list
            while (priorityQueue.Count > 0)
            {
                ListNode node = priorityQueue.Dequeue();
                current.next = new ListNode(node.val);
                current = current.next;

                if (node.next != null)
                {
                    priorityQueue.Enqueue(node.next, node.next.val);
                }
            }
            return dummy.next;
        }

        // This method solves the problem and is called from Program.cs
        public static void Solve()
        {
            List<ListNode> linkedLists = InputLinkedLists();
            
            // Merge k sorted lists
            ListNode result = MergeKSortedLists(linkedLists);

            // Print the merged list
            Console.WriteLine("Merged List:");
            PrintList(result);
        }

        public static ListNode CreateLinkedList(List<int> input)
        {
            if (input.Count == 0)
                return null;

            ListNode head = new ListNode(input[0]);
            ListNode current = head;

            for (int i = 1; i < input.Count; i++)
            {
                current.next = new ListNode(input[i]);
                current = current.next;
            }

            return head;
        }

        // Helper method to print the list
        public static void PrintList(ListNode head)
        {
            ListNode current = head;
            while (current != null)
            {
                Console.Write(current.val + " ");
                current = current.next;
            }
            Console.WriteLine();
        }

        public static List<ListNode> InputLinkedLists()
        {
            Console.WriteLine("Enter a list of lists (e.g. [[1,4,5],[1,3,4],[2,6]]): ");
            string inputStr = Console.ReadLine();
            // Convert input string to list of lists using JSON deserialization
            List<List<int>> lists = JsonConvert.DeserializeObject<List<List<int>>>(inputStr);

            // Convert each list of integers into a linked list
            List<ListNode> linkedLists = new List<ListNode>();
            foreach (var lst in lists)
            {
                linkedLists.Add(CreateLinkedList(lst));
            }

            return linkedLists;
        }
    }
}

File: python_solution/problem23.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Sep 18 12:24:37 2024

@author: sharminalina
"""

from typing import Optional
import heapq

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Problem23:
    
    def mergeKsortedLists(self, linked_lists: list[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = []
    
        # Initialize the heap with the head of each linked list
        for i, lst in enumerate(linked_lists):
            if lst:
                heapq.heappush(min_heap, (lst.val, i, lst))  # Push (value, index, node)
    
        dummy = ListNode()  # Dummy node to simplify handling the head of the result list
        current = dummy
    
        while min_heap:
        # Extract the node with the smallest value
            val, i, node = heapq.heappop(min_heap)
        
        # Attach it to the result linked list
            current.next = ListNode(val)
            current = current.next
            
        # If the extracted node has a next node, push it to the heap
            if node.next:
                heapq.heappush(min_heap, (node.next.val, i, node.next))
    
        return dummy.next
        
    
    
    def list_to_linked_list(self,lst):
        dummy = ListNode()  # A dummy node to simplify handling the head
        current = dummy
        for val in lst:
            current.next = ListNode(val)
            current = current.next
        return dummy.next

    def printList(self, head: ListNode) -> None:
        current = head
        while current is not None:
            print(current.val, end=" ")
            current = current.next
        print()
    
    def input_linked_lists(self):
        input_str = input("Enter a list of lists (e.g. [[1,4,5],[1,3,4],[2,6]]): ")
        lists = eval(input_str)  # Convert string input to a list of lists
        linked_lists = [self.list_to_linked_list(lst) for lst in lists]
        return linked_lists

            



def main():
    print("Problem No. 23 : Marge K sorted List")
    
    
    solution_prob23 = Problem23()
    linked_lists = solution_prob23.input_linked_lists()
   
    result = solution_prob23.mergeKsortedLists(linked_lists);

    # Print the modified list
    print("Marge K sorted List:");
    solution_prob23.printList(result)

if __name__ == "__main__":
    main()